### ISSUE TITLE
User namespaces - Phase 2

### ISSUE THREAD

**Issue by merlinnot** [docker/docker#28593]
I've found few references to this issue:
- #15187 
- [at linux foundation.com](https://events.linuxfoundation.org/sites/events/files/slides/User%20Namespaces%20-%20ContainerCon%202015%20-%2016-9-final_0.pdf)
- [and at this blog.](https://integratedcode.us/2015/10/13/user-namespaces-have-arrived-in-docker/)

 but there's no ticket regarding this problem, AFAIK.

It would be more than useful to create per-container user namespaces (remapped root UID/GID) in one of my projects. I haven't even found any workaround, creating or even changing {U,G}ID doesn't resolve my issues.

Expected behaviour might be sth like
`docker run -map-uid=$HOSTUID:$CONTAINERUID`
`docker run -map-gid=$HOSTGID:$CONTAINERGID`

**Comment by thaJeztah** [comment]
ping @estesp

**Comment by rhatdan** [comment]
@thaJeztah We have been talking about this same issue internally.  I think it would be helpful to agreeing on the CLI even though the kernel/file system is not their yet to make this work.

I would suggest that this is a security thing, so 

docker run --security-opt mapping=uid:$HOSTUID:$CONTAINERUID  --security-opt maping=gid:$HOSTGID:$CONTAINERGIF

**Comment by mheon** [comment]
To elaborate on the kernel/FS work:

Right now, for a single user namespace, images are chown'd as they are pulled to make them match the UID ranges that are mapped into the user namespace (as an example, we might map UIDs 100000-165535 to UID 0-65536 in the container, so all images pulled down will be chowned so root is 100000, not 0, and so on). This approach doesn't really work with multiple user namespaces, because now we might need an image to be 100000-165536 for one user namespace and 200000-265535 for another, meaning we'd need two copies of the same image on disk with different chowns. I did some experimental work into this area, and it's very messy code-wise - you end up breaking storage deduplication in a lot of places, you potentially need multiple image stores, it's not pretty.

There is work going on in the upstream kernel to add a new filesystem called ShiftFS. ShiftFS shifts the UIDs and GIDs of an underlying filesystem, which can be used to eliminate the need for chowning images. Instead of doing a chown when pulling an image down, instead just use ShiftFS to shift the UIDs to match what the user namespace expects. It's a much more elegant and scalable solution, but it's not in the kernel yet (though there is an old version of the patchset posted to the LKML).

**Comment by merlinnot** [comment]
I guess we'll have to wait a little bit longer according to [this](https://lwn.net/Articles/700422/). Last patch for ShifFS was [posted](https://patchwork.kernel.org/project/linux-fsdevel/list/?q=shiftfs) June 1st... I can't find much more information regarding this topic.

I'll try to resolve my issues using bindfs/nfs for now.

@rhatdan It's not only security thing, at least in my case it is not. Currently there's a `--userns` flag which allows you to disable namespaces for specific container, which is actually misleading IMO... I need one-to-one mapping, though most similar solutions define some range instead of single mapping. It makes a little more sense actually.

Please update this thread with some news from the kernel world, as appropriate solution is needed.

**Comment by cpuguy83** [comment]
I'd expect an option to say "all containers get their own user namespace mapping" and then provide a way to tweak that per container (the mapping/sharing).

**Comment by merlinnot** [comment]
Basic mapping is already (implemented)[https://success.docker.com/Datacenter/Apply/Introduction_to_User_Namespaces_in_Docker_Engine].

There is much more to do tho.

**Comment by rhatdan** [comment]
@cpuguy83 Yes this should be a daemon option(s).  But the syntax to override at the docker create point is where this discussion was heading.

I also find it worrysome that the shiftfs changes have stalled as of last summer.  My understanding though is that Eric Biederman has gotten lots of changes into VFS layer of kernel for user namespace that were required to move something like shiftfs along.  I think that shiftfs should be a readonly file system and container runtimes should mount OverlayFS on top to create a writable directory. Having a writable shiftfs seems very dangerous from a security point of view.  

When I played with it, I created a SUID app in my user namespace, and shiftfs created the SUID file as owned by root on the host.

**Comment by merlinnot** [comment]
I've tried to reach out James Bottomley @jejb via Twitter with no success. He's the only person I can find who might know sth in the subject.

Is there some mailing list / IRC worth trying?

**Comment by jejb** [comment]
On Mon, 2016-11-21 at 07:55 -0800, Natan Sągol wrote:
> I've tried to reach out James Bottomley @jejb via Twitter with no
> success.

That would be because twitter is basically broken:  it's email
notifications are useless because it didn't even tell me there was
anything waiting and I don't usually log on unless I have something to
post.

>  He's the only person I can find who might know sth in the subject.

Phil Estes is doing the docker work, so he'll be a good resource

> Is there some mailing list / IRC worth trying?

The original one the patch was on (linux-fsdevel)?

http://marc.info/?l=linux-fsdevel&m=146474103718788

The status is I need to merge with the new superblock user namespace
before doing a repost.

James

**Comment by estesp** [comment]
Thanks James :) Was just about to post that I had heard from you on Friday re: your talk with Eric B. and where you were on re-work of the shiftfs work.

To everyone else on this thread, I think all the requisite parts have been discussed in prior comments, and we (working on upstream Docker and other upstream communities) are all on the same page--we need some filesystem support upstream to really implement this the right way. Everything else requires multiple local caches of layers with various `chown`-ing to get things right, and it gets messy and loses all the advantages of having CoW layered filesystems shared on a container host/node.

How the actual interface looks is obviously TBD, but to have per-container mappings, there must be some specification that you can pass to `docker run`; I agree with @rhatdan it should probably tie in to the existing `--security-opt` infrastructure for seccomp, apparmor, etc., but we can get other opinions as well.

Personally, I think it might be reasonable to leave the daemon option as well for those who are not running multi-tenant and just want the generic use of user namespaces across all containers on that host node. Options to `docker run` would simply override any existing use of user namespace mappings that are daemon-wide.

**Comment by mheon** [comment]
I tend to disagree on tying into `--security-opt` given that we already have a flag for user namespaces. To escape a user namespace, you pass `--userns=host`, so it would seem to make sense to override the default mappings through that flag as well. Format for the flag probably needs some discussion (user and group mappings mean you need to cram a lot of info into the CLI), but discussion on that can probably wait until the changes are ready for a PR.

**Comment by merlinnot** [comment]
Thanks @jejb 👍

**Comment by kidambisrinivas** [comment]
I hit the same problem as well and created a stackoverflow qsn here:

http://stackoverflow.com/questions/40124560/docker-user-namespacing-map-user-in-container-to-host

Please let me know if there is any current workaround for this

**Comment by merlinnot** [comment]
As for now, there's none.

**Comment by nathanjsweet** [comment]
I'd like to ask if it would be worthwhile or doable to implement this feature by allowing the user to opt in at the daemon level to having master copies of the layers and copying and chowning them (as needed) when a new container needs to be spun up. A sort of per user namespace directory of all the layers needed per user namespace. Sure, it sucks that there are no longer any benefits to sharing the layers in such a system, but at least the feature could be implement, and it would certainly go a long way to helping further sandbox docker containers to get this done, rather than wait for a kernel patch to land and slowly trickle into the distributions a few years from now.

Honestly the space tradeoffs aren't that bad (for example, a host with 20 containers all running full ubuntu in different namespaces would only waste 3.5 GB) and in production systems people tend to roll more compact images anyways.

What are people's thoughts?

**Comment by adelton** [comment]
I understand that the problem with per-container user namespace is the need for multiple copies of the images. However, let's assume the daemon itself would be running with uid 808080 for containerized root's uid 0. The pulled images would be stored with most inodes being owned by uid 808080.

If a container would then be started with root's uid being 909090, would it be possible to map both the 808080 and 909090 to 0 in the container, with 909091 being 1, etc? That way the read-only pulled image (808080 on host) would be exposed to the container correctly as uid 0, and so would any changes written to overlay as uid 909090. Can two different host uids be mapped to the same namespaced uid 0?

**Comment by estesp** [comment]
@nathanjsweet I'm actually looking at doing a small proof-of-concept with containerd using the `WithRemappedSnapshot` option in the containerd client library. Once there is deeper integration between moby's use of containerd 1.0 (e.g. using the snapshot drivers) then maybe that implementation could be quite easy. Right now it is actually pretty invasive to try and implement that per-container layer copy for the user namespaces case, which is why you haven't seen much activity there. I agree that for someone who prioritizes that feature above disk space use, it may be a practical middle ground until a more elegant solution is available.

@adelton the premise of having two different host IDs map to container ID 0 (container root) doesn't work, so the rest of that potential workaround is unavailable today. You could have some very unique `subuid`/`subgid` maps that re-use a single root for a set of users; all with different mappings for uid 1 through *{top of range}*, but that really leaves you with very little benefit as most containers will run as the remapped root (808080 in your example). The fact that uid 1 is 909091 in container B and 999991 in container C, and so on, is very little added isolation benefit as the container root will be the same across all containers, not to mention any files **not** owned by root will be in a very weird predicament as far as how they are uid/gid mapped in other containers that have a unique 1-max mapping. Effectively they will all be "unmapped" to the "no such user/group" 65536 id, and will have varying levels of usefulness in that unmapped state.

**Comment by adelton** [comment]
@estep, thanks for the explanation, appreciated.

**Comment by nathanjsweet** [comment]
@estesp what's your timeline on this POC?

**Comment by rijnhard** [comment]
somewhat related, I'd like to throw a spanner in the works if I may. 

Another use case for remapping users/permissions is when using bind mounts. 
In a development environment, the number of times I've had issues bind mounting to a local code directory and then not having the correct permissions/ownership inside the container has happened way more than I care to remember. (And then the endless tunnel of similar issues and resolutions that just don't quite fix the issue).

I have one container with a service in that won't read the files unless it's exactly `-rw-r--r--` and in my code base its `-rwxrwxrwx` (obviously so that I can edit it easily). Grey hair material (if I had any). 

Remapping the ownership and permissions mask in a bind mount would fix a lot of these issues.

**Comment by cyphar** [comment]
It should be noted that with the existence of reflinks it is now possible to `chown` entire images with very little duplication overhead (only the inode metadata is copied -- though this does unfortunately include the xattrs). While this would lead to inode exhaustion in the long-run, it would be a useful stop-gap until we have "shiftfs" or remapped bind-mounts.

**Comment by lucasbasquerotto** [comment]
It seems to me that solving this issue will also solve this (very requested) issue: https://github.com/moby/moby/issues/22415

**Comment by rata** [comment]
Hi! We are adding support for [user namespaces in Kubernetes](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces), we are currently writing patches for containerd and CRIO.

It would be great if dockershim can implement support for it too. However, it seems this issue is a pre-requisite for that, as we need to use different ID mappigs for different pods.

I've created [this issue in dockershim](https://github.com/Mirantis/cri-dockerd/issues/74) but wanted to ask here if there are any plans to fix this issue :)

**Comment by cyphar** [comment]
idmapped mounts have been available in upstream Linux for a few releases now, which means it should be entirely possible for Docker to support this without needing any workarounds (fuse-overlayfs, reflinks, etc). runc doesn't yet support specifying id mappings for mount options but that shouldn't be needed for the Docker rootfs (which Docker mounts itself anyway).

**Comment by rata** [comment]
@cyphar well, we are moving only with Phase I of the KEP now, that doesn't need volumes (we just support the simple ones, that have the same lifecycle of the pod and files are created by the kubelet, like secrets, configmaps, etc.). So, we just need docker to have the ability to run containers with different mappings for now. The simple volumes supported will have the proper permissions for the user in the userns to access them, nothing in the container runtime to do really.

So, what docker needs to solve for this first phase is even simpler: being able to run pods with userns and different uid/gid mapping each that is not possible today, and the making the rootfs accessible to the pod (it can be with idmap mounts as you mentioned, though overlayfs idmap is supported since 5.19+. Or using plain overlayfs, chown and using metacopy param to reduce a little bit the overhead, although need to be careful running out of inodes).

**Comment by cyphar** [comment]
@rata The main justification given why Docker did not implement that feature in "phase 1" of user namespaces was that there wasn't a solution for using the same downloaded and unpacked image for more than one container with different mappings (LXD implemented this by remapping the rootfs for each container if bindfs was not available, but this was a tradeoff between storage usage and security), hence everything using the same mapping by default -- so this was a blocker for Docker for a long time (as you mentioned, there is more than one solution but idmapped mounts are the nicest and most modern). The fact you can support volumes using the same mechanism is a nice bonus.

**Comment by luochen1990** [comment]
This feature is very useful!  Any progress here?

**Comment by rata** [comment]
@luochen1990 I opened this issue in cri-dockerd some years ago, I didn't get any progress to happen: https://github.com/Mirantis/cri-dockerd/issues/74
